snark/src/lib.rs

```diff
@@ -72,26 +66,23 @@ pub trait SNARK<F: PrimeField> {
     /// specified as R1CS constraints.
     fn verify_with_processed_vk(
         circuit_pvk: &Self::ProcessedVerifyingKey,
-        public_input: &[F],
+        instance: &R::Instance,
         proof: &Self::Proof,
     ) -> Result<bool, Self::Error>;
 }
 
 /// A SNARK with (only) circuit-specific setup.
-pub trait CircuitSpecificSetupSNARK<F: PrimeField>: SNARK<F> {
-    /// The setup algorithm for circuit-specific SNARKs. By default, this
-    /// just invokes `<Self as SNARK<F>>::circuit_specific_setup(...)`.
-    fn setup<C: ConstraintSynthesizer<F>, R: RngCore + CryptoRng>(
-        circuit: C,
-        rng: &mut R,
-    ) -> Result<(Self::ProvingKey, Self::VerifyingKey), Self::Error> {
-        <Self as SNARK<F>>::circuit_specific_setup(circuit, rng)
-    }
+pub trait CircuitSpecificSetupSNARK<R: NPRelation>: SNARK<R> {
+    /// The setup algorithm for circuit-specific SNARKs.
+    fn setup<Rng: RngCore + CryptoRng>(
+        index: &R::Index,
+        rng: &mut Rng,
+    ) -> Result<(Self::ProvingKey, Self::VerifyingKey), Self::Error>;
 }
 
 /// A helper type for universal-setup SNARKs, which must infer their computation
 /// size bounds.
-pub enum UniversalSetupIndexError<Bound, E> {
+pub enum IndexingError<Bound, E> {
```

Comment:
Indexing is an operation that happens only for `UniversalSetupSNARK`s, so we don't need to specify that separately here.